util/paused/paused_test.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package paused implements paused helper functions.
+package paused
+
+import (
+	"context"
+	"testing"
+
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	v1beta2conditions "sigs.k8s.io/cluster-api/util/conditions/v1beta2"
+)
+
+func TestEnsurePausedCondition(t *testing.T) {
+	g := NewWithT(t)
+
+	scheme := runtime.NewScheme()
+	g.Expect(clusterv1.AddToScheme(scheme)).To(Succeed())
+
+	normalCluster := &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "some-cluster",
+			Namespace: "default",
+		},
+	}
+	normalClusterWithCondition := normalCluster.DeepCopy()
+	v1beta2conditions.Set(normalClusterWithCondition, pausedCondition(normalClusterWithCondition, normalClusterWithCondition, clusterv1.PausedV1Beta2Condition))
+
+	pausedSpecCluster := normalCluster.DeepCopy()
+	pausedSpecCluster.Spec.Paused = true
+	pausedSpecClusterWithCondition := pausedSpecCluster.DeepCopy()
+	v1beta2conditions.Set(pausedSpecClusterWithCondition, pausedCondition(pausedSpecClusterWithCondition, pausedSpecClusterWithCondition, clusterv1.PausedV1Beta2Condition))
+
+	pausedAnnotationCluster := normalCluster.DeepCopy()
+	pausedAnnotationCluster.SetAnnotations(map[string]string{clusterv1.PausedAnnotation: "true"})
+	pausedAnnotationClusterWithCondition := pausedAnnotationCluster.DeepCopy()
+	v1beta2conditions.Set(pausedAnnotationClusterWithCondition, pausedCondition(pausedAnnotationClusterWithCondition, pausedAnnotationClusterWithCondition, clusterv1.PausedV1Beta2Condition))
+
+	pausedSpecAnnotationCluster := pausedSpecCluster.DeepCopy()
+	pausedSpecAnnotationCluster.SetAnnotations(map[string]string{clusterv1.PausedAnnotation: "true"})
+	pausedSpecAnnotationClusterWithCondition := pausedSpecAnnotationCluster.DeepCopy()
+	v1beta2conditions.Set(pausedSpecAnnotationClusterWithCondition, pausedCondition(pausedSpecAnnotationClusterWithCondition, pausedSpecAnnotationClusterWithCondition, clusterv1.PausedV1Beta2Condition))
+
+	tests := []struct {
```

Comment:
What about adding a few test cases for when the object is not a cluster (or simply use something different than the cluster for the existing ones, object == cluster is an "edge" case)